Source/WebKit/UIProcess/Extensions/WebExtension.cpp

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (C) 2024 Igalia S.L. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WebExtension.h"
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+namespace WebKit {
+
+static constexpr auto manifestVersionManifestKey = "manifest_version"_s;
+
+static constexpr auto nameManifestKey = "name"_s;
+static constexpr auto shortNameManifestKey = "short_name"_s;
+static constexpr auto versionManifestKey = "version"_s;
+static constexpr auto versionNameManifestKey = "version_name"_s;
+static constexpr auto descriptionManifestKey = "description"_s;
+
+bool WebExtension::manifestParsedSuccessfully()
+{
+    if (m_parsedManifest)
+        return !!m_manifest;
+
+    // If we haven't parsed yet, trigger a parse by calling the getter.
+    auto isManifestValid = !!manifest();
+
+    // Validate that the manifest exists as a JSON Object
+    if (!m_manifestJSON->asObject()) {
+        createError(Error::InvalidManifest);
+        m_manifestJSON = JSON::Value::null();
+    }
+
+    return isManifestValid;
```

Comment:
The `asObject()` check should be in the `parseManifest()` method. I moved it in my suggestion. The `createError()` call was also missing a call to `recordError()`.

```suggestion
    // If we haven't parsed yet, trigger a parse by calling the getter.
    return !!manifestObject();
```

Also adopts `manifestObject()` from my other suggestion, to move away from the `NSDictionary` version of `manifest()`.

Reply:
by doing this, the NSDictionary copy of manifest is not parsed, since manifest() is never called (except in some error code it looks like).

Reply:
also by not calling manifest(), m_parsedManifest is never set, which means manifestParsedSuccessfully will call manifestObject, which will continue to call manifestParsedSuccessfully in a loop.

Reply:
Good point. Yeah checking both is fine!